plugins/org.eclipse.n4js.semver/src/org/eclipse/n4js/semver/SEMVERMatcher.java

```diff
@@ -0,0 +1,375 @@
+	 *         {@link #matches(VersionNumber, VersionRangeSet)}
+	 */
+	static public boolean canComputeMatch(VersionNumber proband, NPMVersion npmv) {
+		if (proband == null) {
+			return false;
+		}
+		if (npmv instanceof VersionRangeSet) {
+			return true;
+		}
+		if (npmv instanceof URLVersion) {
+			URLVersion urlVersion = (URLVersion) npmv;
+			return urlVersion.hasSimpleVersion();
+		}
+
+		return false;
+	}
+
+	/**
+	 * This method checks {@link NPMVersion}s whether they match or not. In case the given npm constraint is a SEMVER
+	 * version (i.e. {@link VersionRangeSet} or other subtypes that contain {@link SimpleVersion}), the given proband is
+	 * checked against it and the result of this check is returned. Otherwise {@code true} is returned.
+	 *
+	 * @param proband
+	 *            version that is checked to match the constraint
+	 * @param constraint
+	 *            npm version that may contain a SEMVER version range set or simple version
+	 * @return true iff either the given {@code proband} version matches the given {@code constraint} version or iff the
+	 *         {@code constraint} does neither contain a SEMVER version range nor a simple version
+	 */
+	static public boolean matches(VersionNumber proband, NPMVersion constraint) {
+		if (proband == null) {
+			return false;
+		}
+		if (constraint instanceof VersionRangeSet) {
+			return matches(proband, (VersionRangeSet) constraint);
+		}
+		if (constraint instanceof URLVersion) {
+			return matches(proband, (URLVersion) constraint);
+		}
+		return true;
+	}
+
+	/**
+	 * Compares two SEMVER {@link VersionNumber}s A and B.
+	 * <p>
+	 * In case the argument {@code a} has a pre-release tag, this method returns true only if the given argument
+	 * {@code limit} also has a pre-release tag.
+	 * <p>
+	 * Note that this function cannot cover cases when one version is checked against multiple other versions. In these
+	 * cases the method {@link #matches(VersionNumber, VersionRangeSet)} should be used.
+	 *
+	 * @return relation between two given {@link VersionNumber}s
+	 */
+	static public VersionNumberRelation relation(VersionNumber a, VersionNumber limit) {
+		return relation(a, limit, false);
+	}
+
+	/**
+	 * Compares two SEMVER {@link VersionNumber}s A and B.
+	 * <ul>
+	 * <li/>Returns 0 iff A and B are equal.
+	 * <li/>Returns 1 iff A is greater than B.
+	 * <li/>Returns -1 iff A is smaller than B.
+	 * <li/>Returns -10 otherwise.
+	 * </ul>
+	 * Note that this function cannot cover cases when one version is checked against multiple other versions. In these
+	 * cases the method {@link #matches(VersionNumber, VersionRangeSet)} should be used.
+	 */
+	static public int compare(VersionNumber a, VersionNumber b) {
+		VersionNumberRelation versionRelation = relation(a, b, false);
+		switch (versionRelation) {
+		case Equal:
+			return 0;
+		case Greater:
+			return 1;
+		case Smaller:
+			return -1;
+		case Unrelated:
+			return -10;
```

Comment:
could this cause trouble on client side, e.g. when using `VersionNumber`s in a TreeSet, etc.?

Reply:
I refactored this. Now we have two kinds of compare methods which solve this issue.